src/apple/alohalytics_objc.mm

```
@@ -0,0 +1,354 @@
+      ASIdentifierManager * manager = [ASIdentifierManager sharedManager];
+      info.emplace("isAdvertisingTrackingEnabled", manager.isAdvertisingTrackingEnabled ? "YES" : "NO");
+      if (manager.advertisingIdentifier) {
+        info.emplace("advertisingIdentifier", ToStdString(manager.advertisingIdentifier.UUIDString));
+      }
+    }
+  }
+  if (!info.empty()) {
+    instance.LogEvent("$iosDeviceIds", info);
+  }
+  // Force uploading to get first-time install information before uninstall.
+  instance.Upload();
+}
+
+// Returns <unique id, true if it's the very-first app launch>.
+static std::pair<std::string, bool> InstallationId() {
+  bool firstLaunch = false;
+  NSUserDefaults * userDataBase = [NSUserDefaults standardUserDefaults];
+  NSString * installationId = [userDataBase objectForKey:@"AlohalyticsInstallationId"];
+  if (installationId == nil) {
+    CFUUIDRef uuid = CFUUIDCreate(kCFAllocatorDefault);
+    // All iOS IDs start with I:
+    installationId = [@"I:" stringByAppendingString:(NSString *)CFBridgingRelease(CFUUIDCreateString(kCFAllocatorDefault, uuid))];
+    CFRelease(uuid);
+    [userDataBase setValue:installationId forKey:@"AlohalyticsInstallationId"];
+    [userDataBase synchronize];
+    firstLaunch = true;
+  }
+  return std::make_pair([installationId UTF8String], firstLaunch);
+}
+
+// Returns path to store statistics files.
+static std::string StoragePath() {
+  // Store files in special directory which is not backed up automatically.
+  NSArray * paths = NSSearchPathForDirectoriesInDomains(NSApplicationSupportDirectory, NSUserDomainMask, YES);
+  NSString * directory = [[paths firstObject] stringByAppendingString:@"/Alohalytics/"];
+  NSFileManager * fm = [NSFileManager defaultManager];
+  if (![fm fileExistsAtPath:directory]) {
+    if (![fm createDirectoryAtPath:directory withIntermediateDirectories:YES attributes:nil error:nil]) {
+      NSLog(@"Alohalytics ERROR: Can't create directory %@.", directory);
```

Right. Even if you can't get any statistics about the product, it should remain usable for the user.
This is the main reason why I prefer to silence all exceptions and critical errors in the statistics core.